File: Editorial/0109_Convert_Sorted_List_to_Binary_Search_Tree/C/1_Recursion.c
struct ListNode* FindMiddle(struct ListNode* head) {
    struct ListNode* prev = NULL;
    struct ListNode* slow = head;
    struct ListNode* fast = head;
    while (fast != NULL && fast->next != NULL) {
        prev = slow;
        slow = slow->next;
        fast = fast->next->next;
    }
    if (prev != NULL) prev->next = NULL;
    return slow;
}
struct TreeNode* sortedListToBST(struct ListNode* head) {
    if (head == NULL) return NULL;
    struct ListNode* mid = FindMiddle(head);
    struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val = mid->val;
    if (head == mid) {
        node->left = NULL;
        node->right = NULL;
        return node;
    }
    node->left = sortedListToBST(head);
    node->right = sortedListToBST(mid->next);
    return node;
}

File: Editorial/0109_Convert_Sorted_List_to_Binary_Search_Tree/TypeScript/1_Recursion.ts
/**
 * Definition for singly-linked list.
 * class ListNode { constructor(public val: number, public next: ListNode | null = null) { } }
 *
 * Definition for a binary tree node.
 * class TreeNode { constructor(public val: number, public left: TreeNode | null = null, public right: TreeNode | null = null) { } }
 */
function sortedListToBST(head: ListNode | null): TreeNode | null {
    if (!head) {
        return null;
    }
    const mid = findMiddleElement(head);
    const node: TreeNode = new TreeNode(mid.val);
    if (head === mid) {
        return node;
    }
    node.left = sortedListToBST(head);
    node.right = sortedListToBST(mid.next);
    return node;
}
function findMiddleElement(head: ListNode | null): ListNode | null {
    let prevPtr: ListNode | null = null;
    let slowPtr: ListNode | null = head;
    let fastPtr: ListNode | null = head;
    while (fastPtr !== null && fastPtr.next !== null) {
        prevPtr = slowPtr;
        slowPtr = slowPtr !== null ? slowPtr.next : null;
        fastPtr =
            fastPtr !== null && fastPtr.next !== null
                ? fastPtr.next.next
                : null;
    }
    if (prevPtr !== null) {
        prevPtr.next = null;
    }
    return slowPtr;
}
